packages/langium/src/grammar/langium-grammar-validator.ts

```diff
@@ -96,33 +106,71 @@ export class LangiumGrammarValidator {
         }
     }
 
-    checkGrammarHiddenTokens(grammar: Grammar, accept: ValidationAcceptor): void {
-        if (grammar.hiddenTokens && grammar.hiddenTokens.length > 0) {
-            for (let i = 0; i < grammar.hiddenTokens.length; i++) {
-                const hiddenToken = grammar.hiddenTokens[i];
-                if (!hiddenToken.ref) {
-                    continue;
-                }
-                if (isTerminalRule(hiddenToken.ref)) {
-                    if (hiddenToken.ref.fragment) {
-                        accept('error', 'Cannot use terminal fragments as hidden tokens.', { node: grammar, property: 'hiddenTokens', index: i });
-                    }
-                } else if (hiddenToken.ref) {
-                    accept('error', 'Only terminal rules may be used as hidden tokens.', { node: grammar, property: 'hiddenTokens', index: i });
+    checkGrammarHiddenTokens(grammar: ast.Grammar, accept: ValidationAcceptor): void {
+        if (grammar.definesHiddenTokens) {
+            accept('error', 'Hidden terminals are declared at the terminal definition.', { node: grammar, property: 'definesHiddenTokens', code: IssueCodes.HiddenGrammarTokens});
+        }
+    }
+
+    checkHiddenTerminalRule(terminalRule: ast.TerminalRule, accept: ValidationAcceptor): void {
+        if (terminalRule.hidden && terminalRule.fragment) {
+            accept('error', 'Cannot use terminal fragments as hidden tokens.', { node: terminalRule, property: 'hidden' });
+        }
+    }
+
+    checkEmptyTerminalRule(terminalRule: ast.TerminalRule, accept: ValidationAcceptor): void {
+        try {
+            const regex = terminalRegex(terminalRule);
+            if (new RegExp(regex).test('')) {
+                accept('error', 'This terminal could match an empty string.', { node: terminalRule, property: 'name' });
+            }
+        } catch {
+            // In case the terminal can't be transformed into a regex, we throw an error
+            // As this indicates unresolved cross references or parser errors, we can ignore this here
+        }
+    }
+
+    checkUsedHiddenTerminalRule(rule: ast.AbstractRule, accept: ValidationAcceptor): void {
+        if (ast.isTerminalRule(rule) && rule.hidden) {
+            return;
+        }
+        streamAllContents(rule).forEach(({ node }) => {
+            if (ast.isRuleCall(node) || ast.isTerminalRuleCall(node)) {
+                const ref = node.rule.ref;
+                if (ast.isTerminalRule(ref) && ref.hidden) {
```

Comment:
Is it maybe better performance-wise to associate the validation with `RuleCall | TerminalRuleCall` and check the containing rule in such a case? This would avoid the additional `streamAllContents` call.